server/block/furnace.go

```diff
@@ -0,0 +1,127 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+	"time"
+)
+
+// Furnace is a utility block used for the smelting of blocks and items.
+type Furnace struct {
+	solid
+	bassDrum
+	*smelter
+
+	// Facing is the direction the furnace is facing.
+	Facing cube.Face
+	// Lit is true if the furnace is lit.
+	Lit bool
+}
+
+// NewFurnace creates a new initialised furnace. The smelter is properly initialised.
+func NewFurnace(face cube.Face) Furnace {
+	return Furnace{
+		Facing:  face,
+		smelter: newSmelter(),
+	}
+}
+
+// Tick is called to check if the furnace should update and start or stop smelting.
+func (f Furnace) Tick(_ int64, pos cube.Pos, w *world.World) {
+	if f.Lit && rand.Float64() <= 0.016 { // Every three or so seconds.
+		w.PlaySound(pos.Vec3Centre(), sound.FurnaceCrackle{})
+	}
+	if lit := f.smelter.tickSmelting(time.Second*10, time.Millisecond*100, f.Lit, func(item.SmeltInfo) bool {
+		return true
+	}); f.Lit != lit {
+		f.Lit = lit
+		w.SetBlock(pos, f, nil)
+	}
+}
+
+// EncodeItem ...
+func (f Furnace) EncodeItem() (name string, meta int16) {
+	return "minecraft:furnace", 0
+}
+
+// EncodeBlock ...
+func (f Furnace) EncodeBlock() (name string, properties map[string]interface{}) {
+	if f.Lit {
+		return "minecraft:lit_furnace", map[string]interface{}{"facing_direction": int32(f.Facing)}
+	}
+	return "minecraft:furnace", map[string]interface{}{"facing_direction": int32(f.Facing)}
+}
+
+// UseOnBlock ...
+func (f Furnace) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, _, used := firstReplaceable(w, pos, face, f)
+	if !used {
+		return false
+	}
+
+	place(w, pos, NewFurnace(user.Facing().Face().Opposite()), user, ctx)
+	return placed(ctx)
+}
+
+// BreakInfo ...
+func (f Furnace) BreakInfo() BreakInfo {
+	return newBreakInfo(3.5, pickaxeHarvestable, pickaxeEffective, oneOf(f))
```

Comment:
From the wiki: "A furnace will also drop all of its contents when broken, including XP from smelting items that were extracted by hoppers."

Currently, furnaces don't drop any of their contents as far as I can tell. They also don't hold the experience they built up. Make sure to add that here. Smokers and blast furnaces also don't do this at the moment, so don't forget those either.

Reply:
This already exists:
https://github.com/df-mc/dragonfly/blob/9c3a5b585f558b4288df435276a58f645759aa0e/server/player/player.go#L1728

Reply:
Any reason not to just re-use the code below that? Just sound like we'll have multiple ways to do the same thing with this.
<img width="1025" alt="image" src="https://user-images.githubusercontent.com/16114089/179357232-d5b4c901-afad-4512-a8a9-d44921a562b3.png">

Reply:
I just pushed a commit which should fix that code to work with stuff like furnaces too.

Reply:
We can, yeah. Removed XPContainer and that code.